demos/extruded_continuity/extruded_continuity.py.rst

```diff
@@ -0,0 +1,148 @@
+towards the interior of the domain. :math:`q` can be interpreted as the
+steady-state distribution of a passive tracer carried by a fluid with velocity
+field :math:`\vec{u}`.
+
+We apply an upwind DG method, as we saw in the previous example.  Denoting the
+upwind value of :math:`q` on interior facets by :math:`\widetilde{q}`, the full
+set of equations are then
+
+.. math::
+
+   -\int_\Omega \! q \vec{u_0} \cdot \nabla \phi \, \mathrm{d} x
+   + \int_{\Gamma_\rlap{\mathrm{ext, outflow}}} \! \phi q \vec{u} \cdot \vec{n}
+   \, \mathrm{d} s
+   + \int_{\Gamma_\mathrm{int}} \! (\phi_+ \vec{u} \cdot \vec{n}_+ +
+     \phi_- \vec{u} \cdot \vec{n}_-) \widetilde{q} \, \mathrm{d} S
+   \quad = \quad
+   -\int_{\Gamma_\rlap{\mathrm{ext, inflow}}} \phi q_\mathrm{in} \vec{u} \cdot
+   \vec{n} \, \mathrm{d} s \quad \forall \ \phi \in V,
+
+We will take the domain :math:`\Omega` to be the cuboid
+:math:`\Omega = [0,1] \times [0,1] \times [0,0.2]`. We will use the uniform
+velocity field :math:`\vec{u} = (0, 0, 1)`. :math:`\Gamma_\mathrm{inflow}`
+is therefore the base of the cuboid, while :math:`\Gamma_\mathrm{outflow}`
+is the top. The four vertical sides can be ignored, since
+:math:`\vec{u} \cdot \vec{n} = 0` on these faces.
+
+We use an *extruded* mesh, where the base mesh is a 20 by 20 unit square,
+divided into triangles, with 10 evenly-spaced vertical layers. This gives
+prism-shaped cells. ::
+
+  from firedrake import *
+  m = UnitSquareMesh(20, 20)
+  mesh = ExtrudedMesh(m, layers=10, layer_height=0.02)
+
+We will use a simple piecewise-constant function space for the unknown scalar
+:math:`q`: ::
+
+  V = FunctionSpace(mesh, "DG", 0)
+
+Our velocity will live in a low-order Raviart-Thomas space. The construction of
+this is more complicated than element spaces that have appeared previously. The
+horizontal and vertical components of the field are specified separately. They
+are combined into a single element which is used to build a FunctionSpace. ::
+
+  # RT1 element on a prism
+  W0_h = FiniteElement("RT", "triangle", 1)
+  W0_v = FiniteElement("DG", "interval", 0)
+  W0 = HDivElement(TensorProductElement(W0_h, W0_v))
+  W1_h = FiniteElement("DG", "triangle", 0)
+  W1_v = FiniteElement("CG", "interval", 1)
+  W1 = HDivElement(TensorProductElement(W1_h, W1_v))
+  W_elt = W0 + W1
+  W = FunctionSpace(mesh, W_elt)
+
+As an aside, since our prescibed velocity is purely in the vertical direction, a
+simpler space would have sufficed: ::
+
+  # Vertical part of RT1 element
+  # W_h = FiniteElement("DG", "triangle", 0)
+  # W_v = FiniteElement("CG", "interval", 1)
+  # W_elt = HDivElement(TensorProductElement(W_h, W_v))
+  # W = FunctionSpace(mesh, W_elt)
+
+Or even: ::
+
+  # Why can't everything in life be this easy?
+  # W = VectorFunctionSpace(mesh, "CG", 1)
+
+Next, we set the prescribed velocity field: ::
+
+  velocity = as_vector((0.0, 0.0, 1.0))
+  u = project(velocity, W)
```

Comment:
For this velocity field you could have just written `u = as_vector((0, 0, 1))`, but I guess again you're trying to also introduce TPEs.